packages/block-editor/src/store/reducer.js

```diff
@@ -202,20 +203,36 @@ function withPersistentBlockChange( reducer ) {
 
 		const isExplicitPersistentChange = action.type === 'MARK_LAST_CHANGE_AS_PERSISTENT';
 
-		if ( state !== nextState || isExplicitPersistentChange ) {
-			// Some state changes should not be considered persistent, namely
-			// those which are not a direct result of user interaction.
-			const isPersistentStateChange = ! IGNORED_ACTION_TYPES.has( action.type );
+		// Defer to previous state value (or default) unless changing or
+		// explicitly marking as persistent.
+		if ( state === nextState && ! isExplicitPersistentChange ) {
+			return {
+				...nextState,
+				isPersistentChange: get( state, [ 'isPersistentChange' ], true ),
+			};
+		}
 
-			nextState = {
+		// Some state changes should not be considered persistent, namely those
+		// which are not a direct result of user interaction.
+		const isIgnoredActionType = IGNORED_ACTION_TYPES.has( action.type );
+		if ( isIgnoredActionType ) {
+			return {
 				...nextState,
-				isPersistentChange: isExplicitPersistentChange || (
-					isPersistentStateChange &&
-					! isUpdatingSameBlockAttribute( action, lastAction )
-				),
+				isPersistentChange: false,
```

Comment:
Any reason, we explicitely set this to false when the action is ignored? I'd have expected the same return value as the previous "if" or maybe just `return nextState`

Reply:
The thought is that since state has changed, we want to explicitly communicate that the change is not one which should be considered as persistent. As I understand it, this is key to what's causing the issue in the first place, where `RECEIVE_BLOCKS` is incurring a change in the blocks state, and without also explicitly marking it as false, it will be surfaced up through `onChange` as a persistent change.

https://github.com/WordPress/gutenberg/blob/05570ff1bf0907ec3a801d78d1578a32ca65565b/packages/block-editor/src/components/provider/index.js#L108-L122

I'm fairly convinced this is what's causing the major instability with E2E tests at the moment, where this undo level is created because Puppeteer is fast enough to start typing within a paragraph before the reusable blocks fetch request completes (where the completion of the fetch request triggers `RECEIVE_BLOCKS` and in turn creates an undo level).

Reply:
Makes sense 👍